Feat/adf 1167/extend fe standards documentation (#59)
oat-sa/taohub-articles: articles/Frontend/Architecture/coding-guide.md

Please configure your IDE or development editor to support :

1.  JavaScript ES5 and ES2015+ style
2.  CSS, SASS, HTML, JSON and Handlebars templates, etc.
3.  [ESLint](https://eslint.org), make sure IDE uses `.eslintrc.js` which imports config from [eslint-config-tao](https://github.com/oat-sa/eslint-config-tao)

> 4.  [Prettier](https://prettier.io/), if `prettier.config.js` exists in repository, configure IDE to use, this file imports config from [prettier-config](https://github.com/oat-sa/prettier-config)

Author: That's good idea, I've added package.json config as an example

> 5.  [EditorConfig](https://editorconfig.org/), is one or the options to configure code formatting using [this configuration](../../resources/frontend/lint/.editorconfig)

Reviewer: ```suggestion
5.  [EditorConfig](https://editorconfig.org/), is one of the options to configure code formatting using [this configuration](../../resources/frontend/lint/.editorconfig)
```

Author: Ok, done

The following coding rules apply on ES2015 code :

- no `use strict` anymore in modules, but it needs to be kept outside ES2015 modules.
- Named function expressions can be replaced by the shorthand object notation `{ method(){ } }`
- Do not use `class`, always prefer composition over inheritance. However there are a few use cases `class` would be allowed, for example to extend DOM prototypes, like `Error` to create new error types.
- Try to use `const` by default (immutable) and `let` only for mutable variables (counters).
- Be careful with destructuring, this can create code difficult to read. A reviewer can ask to rewrite a destructred assignment if the produced code is cryptic.
- Use arrow functions for [lambda](https://en.wikipedia.org/wiki/Anonymous_function) to avoid unnecessary lexical scopes, but avoid them for top-level factories or [pure function](https://en.wikipedia.org/wiki/Pure_function) definitions.
- Use template literals instead of string concatenation.

> - Use destructuring and default parameters for method's `options` parameter : `function({label = "", num = 0} = {}) { }`

Reviewer 2: **chore:** Could you also add the rules enforced by ESLint, like `prefer rest over arguments`? It will help developers and it will avoid surprises.

Author: Nice idea, done.

## Best practice & Patterns

JavaScript is an open language, that let's you write code in very different ways, even in different paradigms, from prototypal object oriented to functional programming. In TAO we've selected some programming paradigms and patterns over others. The goal is to bring some consistency and shared practices across the platform.

### Don't repeat Yourself

> "Every piece of knowledge must have a single, unambiguous, authoritative representation within a system."
> _Andy Hunt, The Pragmatic Programmer_

The simple principle will lead to code easier to maintain. So if you write the same code multiple time, think about abstractions. Abstractions doesn't need to be too high level.

### KISS

> "Sometimes, the elegant implementation is just a function. Not a method. Not a class. Not a framework. Just a function."
> _John Carmack, game developer_

If your module needs to expose a function, then your module can expose only a function, especially when there's no state, no side effect!

If multiple functions serve the same purpose they can be grouped into an object serving multiple and independent _static like_ methods :

```js
//a case module util
return {

>   capitalize: function capitalize(inputString) {},
>   camelToSnake: function camelToSnake(inputString) {},

Reviewer 2: **suggestion:** Use ES6+ notation in code samples
```suggestion
  capitalize(inputString) {},
  camelToSnake(inputString) {},
```

Author: Sure, thanks

```js
//the mixin, a separate behavior you'll add on multiple objects
const assignee = {
  getAssignments() {
    return this.assignments;
  },
  setAssignments(deliveries) {
    this.assignments = assignments;
  },
};
const aUser = {

>   firstName: "john",
>   lastName: "snow",

Reviewer 2: **issue:** please follow the code style in examples 😄 (single quotes only)
**suggestion:**
```suggestion
  firstName: 'john',
  lastName: 'snow',
```

Author: Done

The particularity of this pattern is the scope is shared between the target and the mixin : `this` will be shared.
There are multiple ways to achieve this pattern, including using prototypes.

_When to use it ?_
When mixins are pure methods or stateless by preference. There shouldn't be any strong coupling between the target and the mixin (for example the mixin expects a property to be available in the target).

3. Delegation

> Delegation pattern is composition pattern where component delegates functionality to other module

Reviewer 2: **thought:** You could also mention the Adapter pattern. The API is defined by a façade which is delegating the call to a replaceable adapter containing the actual implementation.

Author: Added adapter pattern description and example

```js
const person = {

>   name: "Carl",

Reviewer 2: ```suggestion
  name: 'Carl',
```

Author: That's my Prettier which doesn't work well for `*md` files. Fixed everywhere

  allowanceLimit: 20,
};

const allowance = {
  substract(amount) {
    if (amount <= this.allowanceLimit) {
      this.allowanceLimit -= amount;
    }
  },
  getAllowanceLimit() {
    return this.allowanceLimit;
  },
};

function delegate(source, methods, provider) {
  methods.forEach((methodName) => {

>     source[methodName] = function () {
>       return provider[methodName].apply(source, arguments);
>     };

Reviewer: ```suggestion
    source[methodName] = function (...args) {
      return provider[methodName].apply(source, args);
    };
```
I think today `arguments` is discouraged.

Reviewer 2: **chore:** Totally agree. We should write the example in modern syntax. The `arguments` variable should be avoided. It is enforced by ESLint. I think we should add the ESLint implicit rules to this doc for better transparency and clarity.

**suggestion:** You could also use arrow function in the example, the lexical this of the delegate is not using it.
```suggestion
    source[methodName] = (...args) => provider[methodName].apply(source, args);
```

Author: Yes, I agree. Fixed that everywhere across the document

  });
}

delegate(person, ["substract", "getAllowanceLimit"], allowance);

person.substract(5);
console.log(person.getAllowanceLimit()); //15

person.substract(10);
console.log(person.getAllowanceLimit()); //5

person.substract(10);
console.log(person.getAllowanceLimit()); //5 (did not substract because of limit)
```

`delegate` function does late binding of `provider` mathods to person objects.
Later we call `delegate` to attach allowance operation from `person` to `allowance` module. Allowance methods then update person state

4. Forwarding

Forwarding design pattern is used to completely forward data and control to other module. This is the good approach to manage shared states and distribute permission for shared state operation.

```js
const person1 = {

>   name: "Bob",

Reviewer 2: **suggestion:**
```suggestion
  name: 'Bob',
```

};

const person2 = {

>   name: "Alice",

Reviewer 2: **suggestion:**
```suggestion
  name: 'Alice',
```

};

const familyWallet = {
  balance: 0,
  earn(amount) {
    this.balance += amount;
  },
  spend(amount) {
    this.balance -= amount;
  },
  getBalance() {
    return this.balance;
  },
};

function forward(person, methods, provider) {
  methods.forEach((methodName) => {
    person[methodName] = function () {

>       return provider[methodName].apply(provider, arguments);

Reviewer: Same comment here.

>     person[methodName] = function () {
>       return provider[methodName].apply(provider, arguments);
>     };

Reviewer 2: **suggestion:**
```suggestion
    person[methodName] = (...args) => provider[methodName](...args);
```

  });
}

forward(person1, ["earn", "spend", "getBalance"], familyWallet);
forward(person2, ["spend", "getBalance"], familyWallet);

person1.earn(30);
console.log(person1.getBalance()); // 30

person2.spend(10);
person1.spend(10);

console.log(person2.getBalance()); // 10
```

`forward` function does late binding of `provider` mathods to person objects.
Later we call `forward` to attach wallet operation from `person1` and `person2` to `familyWallet`.

### Factories

> "The best thing about JavaScript is its implementation of functions. It got almost everything right. But, as you should expect with JavaScript, it didn't get everything right."
> _Douglas Crockford, JavaScript: The Good Parts_

When a module needs to keep a state and hide some implementation details, the factory pattern will be selected.

```js
var countDownFactory = function countDownFactory(config) {
  var currentValue = config.value || 0; //private but accessible through the API

>   var interval = null; //kept private

Reviewer: Can we update this to use `let`/`const`?

>   return {
>     getValue: function getValue() {
>       //expose some internal va
>       return currentValue;
>     },
>     start: function start() {
>       interval = setInterval(function () {
>         currentValue--;
>       }, config.delay);
>     },
>     stop: function stop() {
>       clearInterval(interval);
>     },
>     reset: function reset() {},
>   };

Reviewer 2: **chore:** For consistency, we should use the same ES notation for all examples, and if possible use the modern syntax. The other examples are written in ES6+, so better to apply the same here. And please keep the code style in sync too (4 spaces indent).

**suggestion:**
```suggestion
    return {
        getValue() {
            // expose some internal variable
            return currentValue;
        },
        get value() {
            // we can also use getter/setter notation to expose internal variables
            return currentValue;
        },
        start() {
            interval = setInterval(() => (currentValue--), config.delay);
        },
        stop() {
            clearInterval(interval);
        },
        reset() {},
    };
```

```

TAO provides an implementation, the `core/eventifier` module, it has the following features :

- contextualized to an object (events are scoped)
- support [AOP](https://en.wikipedia.org/wiki/Aspect-oriented_programming) style listening (`before` -> `on` -> `after`)
- support namespaces
- support `Promise` (asynchronous handlers)
- supports context spreading

Please check out the [eventifier documentation](#eventifier).

### Provider

When multiples implementation of a given API can be defined, or dynamically defined, the provider pattern is used.

```js
const jsonDataProvider = {
  requestData() {
    //read some json file and return data
  },
};

const csvDataProvider = {
  requestData() {
    //read some csv file and return data
  },
};

const person = {
  dataProvider,
  data,
  updateData() {
    this.data = this.dataProvider.requestData();
  },
  registerProvider(provider) {
    this.dataProvider = provider;
  },
};
```

> In this example two providers are defined, which have same `requestData` API method. `person` object does not request data directly and doesn't implement any data parsing logic, it just registers the appropriated provider and request data from it. This allows to decouple components, keep components small and easy testable.

Reviewer 2: **chore:** You could also mention it works well together with the delegation pattern.

Author: Added few words on the topic, thanks for idea
